### PR TITLE
feat(cmd/cp): add oci-layout-path flag

### LINKED ISSUE #1505: `oras cp` fails if `org.opencontainers.image.ref.name` contains a full reference
### What happened in your environment?

We use OCI images to package eBPF programs in [Inspektor Gadget](https://github.com/inspektor-gadget/inspektor-gadget/). We have an [`ig image export`](https://www.inspektor-gadget.io/docs/v0.32.0/reference/images#export) command that creates an oci layout file with the image. I wanted to use `oras cp` to copy the image to a remote registry, however it fails with:

```bash 
$ oras cp --from-oci-layout ./mygadget.tar.folder:ghcr.io/inspektor-gadget/gadget/mygadget:latest localhost:5000/foo:latest --debug
Error: invalid argument "./mygadget.tar.folder:ghcr.io/inspektor-gadget/gadget/mygadget:latest": failed to find path "./mygadget.tar.folder:ghcr.io/inspektor-gadget/gadget/mygadget": stat ./mygadget.tar.folder:ghcr.io/inspektor-gadget/gadget/mygadget: no such file or directory
```

The issue seems to be related to the fact that the `org.opencontainers.image.ref.name` annotation is set to `ghcr.io/inspektor-gadget/gadget/mygadget:latest` in index.json:

mygadget.tar.folder/index.json:
```json
{
    "schemaVersion": 2,
    "manifests": [
        {
            "mediaType": "application/vnd.oci.image.manifest.v1+json",
            "digest": "sha256:69fb10d45f448525a39bb26ae2af71e57fe2b7aa774547eb57da38ddf08b12e0",
            "size": 1278,
            "annotations": {
                "org.opencontainers.image.created": "2024-09-19T13:06:53-05:00",
                "org.opencontainers.image.description": "TODO: Fill the gadget description",
                "org.opencontainers.image.documentation": "TODO: Fill the gadget documentation URL",
                "org.opencontainers.image.source": "TODO: Fill the gadget source code URL",
                "org.opencontainers.image.title": "TODO: Fill the gadget name",
                "org.opencontainers.image.url": "TODO: Fill the gadget homepage URL"
            },
            "platform": {
                "architecture": "arm64",
                "os": "linux"
            }
        },
        {
            "mediaType": "application/vnd.oci.image.index.v1+json",
            "digest": "sha256:f31a444623261efcc68601704dd3d980692a73a55f5031c8ddca58cc8a23876e",
            "size": 1816,
            "annotations": {
                "org.opencontainers.image.created": "2024-09-19T13:06:53-05:00",
                "org.opencontainers.image.description": "TODO: Fill the gadget description",
                "org.opencontainers.image.documentation": "TODO: Fill the gadget documentation URL",
                "org.opencontainers.image.ref.name": "ghcr.io/inspektor-gadget/gadget/mygadget:latest",
                "org.opencontainers.image.source": "TODO: Fill the gadget source code URL",
                "org.opencontainers.image.title": "TODO: Fill the gadget name",
                "org.opencontainers.image.url": "TODO: Fill the gadget homepage URL"
            }
        },
        {
            "mediaType": "application/vnd.oci.image.manifest.v1+json",
            "digest": "sha256:f99e163810194bd7fc50e23c9f1d7d827e464aac1766892f3747b1764301c3ae",
            "size": 1278,
            "annotations": {
                "org.opencontainers.image.created": "2024-09-19T13:06:53-05:00",
                "org.opencontainers.image.description": "TODO: Fill the gadget description",
                "org.opencontainers.image.documentation": "TODO: Fill the gadget documentation URL",
                "org.opencontainers.image.source": "TODO: Fill the gadget source code URL",
                "org.opencontainers.image.title": "TODO: Fill the gadget name",
                "org.opencontainers.image.url": "TODO: Fill the gadget homepage URL"
            },
            "platform": {
                "architecture": "amd64",
                "os": "linux"
            }
        }
    ]
}
```

According to https://github.com/opencontainers/image-spec/blob/v1.1.0-rc5/annotations.md, `ghcr.io/inspektor-gadget/gadget/mygadget:latest` is a valid value for the annotation. 

### Possible Fix

The following diff fixes the issue:

```diff
diff --git a/cmd/oras/internal/option/target.go b/cmd/oras/internal/option/target.go
index 489eeb8..33b83af 100644
--- a/cmd/oras/internal/option/target.go
+++ b/cmd/oras/internal/option/target.go
@@ -38,7 +38,6 @@ import (
        "oras.land/oras-go/v2/registry/remote/auth"
        "oras.land/oras-go/v2/registry/remote/errcode"
        oerrors "oras.land/oras/cmd/oras/internal/errors"
-       "oras.land/oras/cmd/oras/internal/fileref"
 )

 const (
@@ -120,6 +119,20 @@ func (opts *Target) Parse(cmd *cobra.Command) error {
        }
 }

+// Parse parses file reference on unix.
+func parseModified(reference string, defaultMetadata string) (filePath, metadata string, err error) {
+       i := strings.Index(reference, ":")
+       if i < 0 {
+               filePath, metadata = reference, defaultMetadata
+       } else {
+               filePath, metadata = reference[:i], reference[i+1:]
+       }
+       if filePath == "" {
+               return "", "", fmt.Errorf("found empty file path in %q", reference)
+       }
+       return filePath, metadata, nil
+}
+
 // parseOCILayoutReference parses the raw in format of <path>[:<tag>|@<digest>]
 func (opts *Target) parseOCILayoutReference() error {
        raw := opts.RawReference
@@ -132,7 +145,7 @@ func (opts *Target) parseOCILayoutReference() error {
        } else {
                // find `tag`
                var err error
-               path, ref, err = fileref.Parse(raw, "")
+               path, ref, err = parseModified(raw, "")
                if err != nil {
                        return errors.Join(err, errdef.ErrInvalidReference)
                }
```

I'm willing to open a fix once we agree on a solution. 

### What did you expect to happen?

`oras cp` should have copied the image to the remote repository

### How can we reproduce it?

Set the `org.opencontainers.image.ref.name` annotation to something like `foo:bar` in index.json and try to run `oras cp --from-oci-layout ./myfolder:foo:bar localhost:5000/foo:latest`

### What is the version of your ORAS CLI?

Manually compiled from 961e9f85b2c5a771f7674513cd1393f2e18e85ea

### What is your OS environment?

Ubuntu 22.04 (it really doesn't matter for this bug)

### Are you willing to submit PRs to fix it?

- [X] Yes, I am willing to fix it.

### PR DESCRIPTION
**Pull request by mauriciovasquezbernal**
**What this PR does / why we need it**:

The oci-layout can't be used when the image reference contains a slash ( see issue 1505). This PR introduces a new oci-layout-path that explicitly receives the path of the oci layout fixing the parsing ambiguity.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1505

**Please check the following list**:
- [X]  Does the affected code have corresponding tests, e.g. unit test, E2E test?
- [X]  Does this change require a documentation update?
- [ ]  Does this introduce breaking changes that would require an announcement or bumping the major version?
- [ ]  Do all new files have an appropriate license header?

### CODE REVIEW

File: cmd/oras/internal/option/target_test.go [hunk 1]
```diff
@@ -78,11 +78,12 @@ func Test_parseOCILayoutReference(t *testing.T) {
 	}{
 		{"Empty input", "", "", "", true},
 		{"Empty path and tag", ":", "", "", true},
-		{"Empty path and digest", "@", "", "", false},
+		{"Empty path and digest", "@", "", "", true},
 		{"Empty digest", "path@", "path", "", false},
 		{"Empty tag", "path:", "path", "", false},
 		{"path and digest", "path@digest", "path", "digest", false},
 		{"path and tag", "path:tag", "path", "tag", false},
+		{"path with full reference", "path:repo:tag", "path", "repo:tag", false},
```
Comment (TerryHowe) [inline cmd/oras/internal/option/target_test.go:86]: Shouldn't the path be `path:repo`?

Comment (sabre1041) [reply cmd/oras/internal/option/target_test.go:86]: `path:repo` fails, but the included content passes

**Review by sabre1041 — APPROVED** [review]
LGTM

File: cmd/oras/internal/option/target.go [hunk 2]
```diff
@@ -123,20 +122,22 @@ func (opts *Target) Parse(cmd *cobra.Command) error {
...
 		path = raw[:idx]
 		ref = raw[idx+1:]
-	} else {
-		// find `tag`
-		var err error
-		path, ref, err = fileref.Parse(raw, "")
-		if err != nil {
-			return errors.Join(err, errdef.ErrInvalidReference)
-		}
+	} else if idx := strings.Index(raw, ":"); idx != -1 {
```
Comment (TerryHowe) [inline cmd/oras/internal/option/target.go:131]:
```suggestion
	} else if idx := strings.LastIndex(raw, ":"); idx != -1 {
```

**Review by TerryHowe — CHANGES_REQUESTED** [review]

**Comment by mauriciovasquezbernal** [comment]
I updated the PR with the --from-oci-layout-path idea discussed on #1505. I'll update the documentation once we agree on this approach.

File: cmd/oras/internal/option/target.go [hunk 3]
```diff
@@ -96,13 +97,21 @@ func (opts *Target) ApplyFlagsWithPrefix(fs *pflag.FlagSet, prefix, description
 
 // Parse gets target options from user input.
 func (opts *Target) Parse(cmd *cobra.Command) error {
+	if opts.IsOCILayout && opts.Path != "" {
+		return fmt.Errorf("flag %q is not supported with %q", "oci-layout-path", "oci-layout")
+	}
```
Comment (shizhMSFT) [inline cmd/oras/internal/option/target.go:102]: Consider `CheckMutuallyExclusiveFlags`. Examples see
https://github.com/oras-project/oras/blob/97cb3766ad2055ddbf3cb5dd49c2725ae0b12148/cmd/oras/internal/option/remote.go#L160-L165

**Review by shizhMSFT — CHANGES_REQUESTED** [review]

File: cmd/oras/internal/option/target.go (hunk 3 above)
Comment (mauriciovasquezbernal) [reply cmd/oras/internal/option/target.go]: Thanks, updated it and its corresponding test.

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Comment by TerryHowe** [comment]
It seems like it would be easy to add a e2e test for this.

**Comment by mauriciovasquezbernal** [comment]
> It seems like it would be easy to add a e2e test for this.

I added some tests. There is some duplication with the existing layout tests but I didn't find an easy way to avoid it.

**Review by TerryHowe — APPROVED** [review]
/lgtm

Thanks!

**Pull request merged**